### PR TITLE
Upgrade to bootstrap 4

### LINKED ISSUE #2819: Update to Bootstrap 4
Update templates from Bootstrap 3 to use Bootstrap 4.

### PR DESCRIPTION
**Pull request by brinehart**
**What issue does this PR address?**
This PR updates the CSS/JS/Views to Bootstrap 4.3.1. 
Resolves #2819.

**Does this PR introduce a breaking change?**
No. All changes are front end only changes.

**Please check if the PR fulfills these requirements**
- [x] The commit follows our [guidelines](https://github.com/IdentityServer/IdentityServer4/blob/master/.github/CONTRIBUTING.md)
- N/A Unit Tests for the changes have been added (for bug fixes / features) 

**Other information**:

### CODE REVIEW

File: src/IdentityServer4/host/Views/Account/Login.cshtml [hunk 1]
```diff
@@ -1,80 +1,76 @@
 @model LoginViewModel
 
 <div class="login-page">
-    <div class="page-header">
-        <h1>Login</h1>
-    </div>
-
-    <partial name="_ValidationSummary" />
+	<div class="page-header">
```
Comment (kri5t) [inline src/IdentityServer4/host/Views/Account/Login.cshtml]: It looks like you are using tabs for indenting here. Should probably be changed back to spaces.

Comment (kri5t) [reply src/IdentityServer4/host/Views/Account/Login.cshtml]: I can see that it is an issue in more than this file.

Comment (brinehart) [reply src/IdentityServer4/host/Views/Account/Login.cshtml]: ah. the old Tabs vs Spaces debate. I'll get those changed over to spaces and push the changes up today.

**Comment by leastprivilege** [comment]
I checked it out manually and ran the build - it fails

`C:\local\identity\server4\IdentityServer4\src\IdentityServer4\host\Views\Account\LoggedOut.cshtml(1,27): error RZ1017: Unexpected literal following the 'model' directive. Expected 'line break'. [C:\local\identity\server4\IdentityServer4\src\IdentityServer4\host\Host.csproj]`

Can you fix that?

**Comment by brinehart** [comment]
> I checked it out manually and ran the build - it fails
> 
> `C:\local\identity\server4\IdentityServer4\src\IdentityServer4\host\Views\Account\LoggedOut.cshtml(1,27): error RZ1017: Unexpected literal following the 'model' directive. Expected 'line break'. [C:\local\identity\server4\IdentityServer4\src\IdentityServer4\host\Host.csproj]`
> 
> Can you fix that?

All fixed. I stupidly had the "Prettier" VS Code extension enabled. It did NOT make the code prettier. . .

**Comment by brockallen** [comment]
> All fixed

Great, thanks! I'll have a look this weekend.

**Comment by brockallen** [comment]
Hey. The consent screen looks a little wonky. Can you have a look, please? 

image

**Comment by SIkebe** [comment]
Bootstrap 3.X moved to EOL in 24 Jul 2019. What is the current status of this PR?
https://blog.getbootstrap.com/2019/07/24/lts-plan/

**Comment by brinehart** [comment]
Sorry I got caught up with a few projects at work. Will try to get to addressing Brock’s feedback this week if possible.
    	
    	

    	Get Outlook for iOS
    
  




On Wed, Aug 21, 2019 at 9:22 PM -0400, "SIkebe" <notifications@github.com> wrote:












Bootstrap 3.X moved to EOL in 24 Jul 2019. What is the current status of this PR?

https://blog.getbootstrap.com/2019/07/24/lts-plan/

**Comment by SIkebe** [comment]
@brinehart Thank you for your great job! Much appreciated.

**Comment by leastprivilege** [comment]
Nothing seems to happen here. Can we close?

**Comment by brinehart** [comment]
Sorry I got busy at work. I think I will be able to address the design concerns this week so it can be used finally.
-- 
 Securely sent with Tutanota. Get your own encrypted, ad-free mailbox: 
 https://tutanota.com


Nov 2, 2019, 02:23 by notifications@github.com:

**Comment by SIkebe** [comment]
@brinehart How's it going? The bot added "wontfix" label, but this PR is very valuable since Bootstrap 3.X ended support last summer.

**Comment by brinehart** [comment]
Any help anyone could offer would be greatly appreciated. I have been unable to get a client working correctly on my local machine in order to test and see this screen (working on a mac). When I take a look at the code I am unsure why it is being rendered the way brockallen's screenshot looks above. According to the View it should be rendering correctly according to the Bootstrap 4 spec unless there is either something that I am missing or  maybe another view I am not seeing that might be doing this?

File: src/IdentityServer4/host/Views/Consent/Index.cshtml [hunk 2]
```diff
@@ -1,82 +1,82 @@
...
+      <partial name="_ValidationSummary" />
+
+      <form asp-action="Index" class="consent-form">
+        <input type="hidden" asp-for="ReturnUrl" />
+
+        <div>Uncheck the permissions you do not wish to grant.</div>
+
+        @if (Model.IdentityScopes.Any())
+        {
+        <div class="card consent-buttons">
```
Comment (vanillajonathan) [inline src/IdentityServer4/host/Views/Consent/Index.cshtml]: Consider adding the `.mb-3` class for a bottom margin.

File: src/IdentityServer4/host/Views/Consent/Index.cshtml [hunk 3]
```diff
@@ -1,82 +1,82 @@
...
+          <div class="card-header">
+            <span class="glyphicon glyphicon-user"></span>
+            Personal Information
+          </div>
+          <ul class="list-group">
```
Comment (vanillajonathan) [inline src/IdentityServer4/host/Views/Consent/Index.cshtml]: Consider adding the `.list-group-flush` class, which may be used for when list groups are used inside cards.

Comment (brinehart) [reply src/IdentityServer4/host/Views/Consent/Index.cshtml]: I have added this. Can you please give it a shot to see if it looks right now?

File: src/IdentityServer4/host/Views/Consent/Index.cshtml (hunk 2 above)
Comment (brinehart) [reply src/IdentityServer4/host/Views/Consent/Index.cshtml]: I have added this. Can you please give it a shot to see if it looks right now?

File: src/IdentityServer4/host/Views/Account/Login.cshtml [hunk 4]
```diff
@@ -1,88 +1,87 @@
...
+                <label asp-for="Password"></label>
+                <input type="password" class="form-control" placeholder="Password" asp-for="Password"
+                  autocomplete="off">
+              </div>
+              @if (Model.AllowRememberLogin)
+              {
+              <div class="form-group login-remember custom-control custom-checkbox">
+                <input type="checkbox" class="custom-control-input" asp-for="RememberLogin" />
+                <label class="custom-control-label" asp-for="RememberLogin">
+                  <strong>Remember My Login</strong>
```
Comment (vanillajonathan) [inline src/IdentityServer4/host/Views/Account/Login.cshtml]: Consider removing the `<strong>` element.

File: src/IdentityServer4/host/Views/Account/Login.cshtml [hunk 5]
```diff
@@ -1,88 +1,87 @@
...
-        {
-            <div class="alert alert-warning">
-                <strong>Invalid login request</strong>
-                There are no login schemes configured for this client.
-            </div>
-        }
+    @if (!Model.EnableLocalLogin && !Model.VisibleExternalProviders.Any())
+    {
+    <div class="alert alert-warning">
+      <strong>Invalid login request</strong>
```
Comment (vanillajonathan) [inline src/IdentityServer4/host/Views/Account/Login.cshtml]: Consider replacing with `<h4 class="alert-heading">Invalid login request</h4>`.

File: src/IdentityServer4/host/Views/Consent/Index.cshtml [hunk 6]
```diff
@@ -1,82 +1,82 @@
...
+        }
+
+        <div class="consent-buttons">
+          <button name="button" value="yes" class="btn btn-primary" autofocus>Yes, Allow</button>
+          <button name="button" value="no" class="btn btn-link">No, Do Not Allow</button>
+          @if (Model.ClientUrl != null)
+          {
+          <a class="float-right btn btn-secondary" target="_blank" href="@Model.ClientUrl">
+            <span class="glyphicon glyphicon-info-sign"></span>
+            <strong>@Model.ClientName</strong>
```
Comment (vanillajonathan) [inline src/IdentityServer4/host/Views/Consent/Index.cshtml]: Consider removing the `<strong>` element.

File: src/IdentityServer4/host/Views/Consent/Index.cshtml [hunk 7]
```diff
@@ -1,82 +1,82 @@
...
+            {
+            <partial name="_ScopeListItem" model="@scope" />
+            }
+          </ul>
+        </div>
+        }
+
+        @if (Model.ResourceScopes.Any())
+        {
+        <div class="card">
```
Comment (vanillajonathan) [inline src/IdentityServer4/host/Views/Consent/Index.cshtml]: Consider adding the `mb-3` class.

File: src/IdentityServer4/host/Views/Device/UserCodeConfirmation.cshtml [hunk 8]
```diff
@@ -1,93 +1,93 @@
...
-                </div>
-            </form>
+        <div class="consent-buttons d-flex">
+          <button name="button" value="yes" class="btn btn-primary px-2" autofocus>Yes, Allow</button>
+          <button name="button" value="no" class="btn btn-link px-2">No, Do Not Allow</button>
+          @if (Model.ClientUrl != null)
+          {
+          <a class="ml-auto px-2 btn btn-default" target="_blank" href="@Model.ClientUrl">
+            <span class="glyphicon glyphicon-info-sign"></span>
+            <strong>@Model.ClientName</strong>
```
Comment (vanillajonathan) [inline src/IdentityServer4/host/Views/Device/UserCodeConfirmation.cshtml]: Consider removing the `<strong>` element.

File: src/IdentityServer4/host/Views/Consent/Index.cshtml [hunk 9]
```diff
@@ -1,82 +1,82 @@
...
+          <div class="card-header">
+            <span class="glyphicon glyphicon-tasks"></span>
+            Application Access
+          </div>
+          <ul class="list-group">
```
Comment (vanillajonathan) [inline src/IdentityServer4/host/Views/Consent/Index.cshtml]: Consider adding the `list-group-flush` class.

File: src/IdentityServer4/host/Views/Device/UserCodeConfirmation.cshtml [hunk 10]
```diff
@@ -1,93 +1,93 @@
...
-                    </div>
-                }
+        @if (Model.IdentityScopes.Any())
+        {
+        <div class="card my-1 consent-buttons">
+          <div class="card-header">
+            <span class="glyphicon glyphicon-user"></span>
+            Personal Information
+          </div>
+          <ul class="list-group">
```
Comment (vanillajonathan) [inline src/IdentityServer4/host/Views/Device/UserCodeConfirmation.cshtml]: Consider adding the `list-group-flush` class.

File: src/IdentityServer4/host/Views/Device/UserCodeConfirmation.cshtml [hunk 11]
```diff
@@ -1,93 +1,93 @@
...
-                    </div>
-                }
+        @if (Model.ResourceScopes.Any())
+        {
+        <div class="card my-2">
+          <div class="card-header">
+            <span class="glyphicon glyphicon-tasks"></span>
+            Application Access
+          </div>
+          <ul class="list-group">
```
Comment (vanillajonathan) [inline src/IdentityServer4/host/Views/Device/UserCodeConfirmation.cshtml]: Consider adding the `list-group-flush` class.

**Comment by vanillajonathan** [comment]
@brinehart I am unable to do that because I don't have a working IdentityServer4 setup.

**Comment by remcok** [comment]
@brockallen Is this PR going to be approved? Hope you can find some time to review because Bootstrap 3.3.5 is really getting ancient...

**Comment by brockallen** [comment]
Yes... we're working on our 4.0 milestone, so it will be looked at.

**Comment by remcok** [comment]
Ok, thanks for the quick reply. Looking forward to 4.0

**Comment by brockallen** [comment]
@brinehart so after almost a year, i finally got around to working on this. I used your branch as my starting point, I did additional re-learning of bootstrap, then i made the other tweaks to get this done. My PR is over here (https://github.com/IdentityServer/IdentityServer4/pull/4427) so we'll use that one. Thank you so much for sticking with this and pestering me to finally get to it. 

Thanks!

**Pull request closed without merging**